File: answers/leetcode/383. Ransom Note/Solution.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        letter_map = dict()
        for letter in magazine:
            letter_map[letter] = letter_map.get(letter, 0) + 1
        for letter in ransomNote:
            if letter in letter_map:
                letter_map[letter] -= 1
            if letter not in letter_map or letter_map[letter] < 0:
                return False
        return True

File: answers/leetcode/383. Ransom Note/Solution.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> notes = new HashMap<Character, Integer>();
        for (char letter : magazine.toCharArray()) {
            if (! notes.containsKey(letter)) {
                notes.put(letter, 0);
            }
            notes.put(letter, notes.get(letter) + 1);
        }
        for (char letter : ransomNote.toCharArray()) {
            if (notes.containsKey(letter)) {
                notes.put(letter, notes.get(letter) - 1);
                if (notes.get(letter) < 0) {
                    return false;
                }
            } else {
                return false;
            }
        }
        return true;
    }
}